### ISSUE TITLE
特征图维度不匹配

### ISSUE THREAD

**Issue by huang229** [xggIoU/centernet_tensorflow_wilderface_voc#8]
嗨，标签的维度是128*128,但darknet53_body返回的尺度是16* 16、32*32、64*64，所以最后特征图是64*64，这就与标签维度不匹配了。谢谢你的git，我的实现好了。

**Comment by xggIoU** [comment]
这个你需要自己改，downsample rate 不一样的，理论上特征图越大越好。

**Comment by huang229** [comment]
训练了pascalvoc2007数据，特征图尺寸128*128， 训练了20多个epoch，损失下不去，
total_loss:42.2115 cls_loss:40.0113 offset_loss:0.6196 size_loss:15.8061 learning_rate:0.000081000
total_loss:13.9295 cls_loss:11.4873 offset_loss:0.5177 size_loss:19.2449 learning_rate:0.000081000
total_loss:10.6922 cls_loss:8.3126 offset_loss:0.4995 size_loss:18.8012 learning_rate:0.000081000
total_loss:10.6922 cls_loss:8.3126 offset_loss:0.4995 size_loss:18.8012 learning_rate:0.000081000
total_loss:19.9303 cls_loss:17.2530 offset_loss:0.4911 size_loss:21.8619 learning_rate:0.000081000
total_loss:20.5976 cls_loss:17.9423 offset_loss:0.5158 size_loss:21.3951 learning_rate:0.000081000

**Comment by xggIoU** [comment]
我的训练了很久才收敛了，而且表现跟单类相比差远了；多分类很难训练，要么就试试强一点的backbone。

**Comment by huang229** [comment]
我认为多分类的类别损失函数设计不合理，造成了负样本比例太重。应该只取有目标的特征图进行损失计算，则参考了maskrcnn，你认为？你训练了多少个epoch才收敛的？

**Comment by xggIoU** [comment]
nice，我觉得你说的有道理，可以试下只算有目标的特征图损失。当时我训练的时候差不多30多epoch收敛的，但是在非训练集上效果有点感人，要么是我没作数据增强，要么就是这个损失不对。

**Comment by huang229** [comment]
你训练的多分类，类别多少种？我20个类别的，60个epoch都没见收敛，我的batch_szie = 8, 好急人。

**Comment by huang229** [comment]
你要是把loss = loss - (pos_loss + neg_loss) / num_pos写成loss = loss - (pos_loss ) / num_pos，你会发现训练输出的中心点值就不会变得很小，反推回去，就是感觉在多分类时，负样本损失不合理。

**Comment by xggIoU** [comment]
我训练的也是voc20类，这个loss函数是从官方代码改过来的，只是从pytorch代码变成了tf代码而已。你把loss改了后效果如何

**Comment by xggIoU** [comment]
loss函数我觉得是应该改成loss-pos_loss/num_pos 毕竟是多分类。之前的loss的写法是二分类交叉熵loss的写法。我觉得问题多半还真是这个原因。

**Comment by huang229** [comment]
我改了损失函数，还是没训练出来，训练了一个晚上，global_step * batch_size /(img_num) = epcoh。
你训练Pascalvoc2007数据时，特征图尺寸多大？
time:3443.1420 global_step:106907 total_loss:4.5916 cls_loss:3.1902 offset_loss:0.4403 size_loss:0.9611 learning_rate:0.000000008
time:3467.8351 global_step:106957 total_loss:4.2589 cls_loss:2.7971 offset_loss:0.4644 size_loss:0.9973 learning_rate:0.000000008

**Comment by xggIoU** [comment]
512/8=64，另外特征图越大，负样本越多，正负比例越大。

**Comment by huang229** [comment]
原文用到了deformable convolution和deformable RoI pooling，所以效果好，你觉得会是这个原因吗？我觉得不是，这两个层论文展示的优势是无论物体怎么形变，提取的特征点都更集中于物体周围。但这不能说明其它目标检测网络作为基网络时，得到的结果是天差地别，就算差，也不至于差太多。

**Comment by xggIoU** [comment]
我也是这个观点，我觉得主要还是损失上，如何优化目标是关键

**Comment by yeyuxmf** [comment]
朋友，原作者的是pytorch0.4实现的，我最近又再看论文，然后把原作者的移植到了pytorch1.2感觉很容易移植，代码几乎没有做任何改动，下载pascalvoc数据集就可以直接训练，自己训练的模型和作者的效果差不多
